File: backend/xer_import.py
"""Primavera P6 XER importer -> internal activity model."""
import re
from datetime import date

TYPE_MAP = {
    "TT_Mile": "Milestone",
    "TT_FinMile": "Milestone",
    "TT_StartMile": "Milestone",
    "TT_WBS": "Summary",
    "TT_LOE": "Summary",
    "TT_Task": "Task",
    "TT_Rsrc": "Task",
}

PRED_MAP = {"PR_FS": "FS", "PR_SS": "SS", "PR_FF": "FF", "PR_SF": "SF"}

CSTR_MAP = {
    "CS_MSO": "SNET", "CS_MSOB": "SNET", "CS_ALAP": "", "CS_MANDSTART": "MSO",
    "CS_MEO": "FNLT", "CS_MEOB": "FNLT", "CS_MANDFIN": "FNLT",
    "CS_MSOA": "SNET", "CS_MEOA": "FNLT",
}


def parse_tables(text: str) -> dict:
    """XER -> {table_name: [ {field: value} ]}."""
    tables, name, fields = {}, None, []
    for raw in text.splitlines():
        if not raw or raw.startswith("ERMHDR"):
            continue
        parts = raw.split("\t")
        tag = parts[0]
        if tag == "%T":
            name = parts[1].strip()
            tables[name] = []
            fields = []
        elif tag == "%F":
            fields = [p.strip() for p in parts[1:]]
        elif tag == "%R" and name and fields:
            values = parts[1:]
            row = {f: (values[i] if i < len(values) else "") for i, f in enumerate(fields)}
            tables[name].append(row)
        elif tag == "%E":
            break
    return tables


def _date(value):
    if not value:
        return None
    m = re.match(r"(\d{4})-(\d{2})-(\d{2})", value.strip())
    if not m:
        return None
    return date(int(m.group(1)), int(m.group(2)), int(m.group(3))).isoformat()


def _hours_to_days(value):
    try:
        return max(0, int(round(float(value or 0) / 8.0)))
    except (TypeError, ValueError):
        return 0


def _wbs_paths(rows):
    """wbs_id -> {'names': [l1, l2, ...], 'code': '1.2'}"""
    by_id = {r["wbs_id"]: r for r in rows}
    children = {}
    for r in rows:
        children.setdefault(r.get("parent_wbs_id") or "", []).append(r)
    for kids in children.values():
        kids.sort(key=lambda r: (int(r.get("seq_num") or 0), r.get("wbs_name", "")))

    out = {}

    def walk(node_id, names, code):
        for i, kid in enumerate(children.get(node_id, []), start=1):
            is_root = kid.get("proj_node_flag") == "Y"
            kid_names = names if is_root else names + [kid.get("wbs_name", "")]
            kid_code = code if is_root else (f"{code}.{i}" if code else str(i))
            out[kid["wbs_id"]] = {"names": kid_names, "code": kid_code}
            walk(kid["wbs_id"], kid_names, kid_code)

    roots = [r for r in rows if (r.get("parent_wbs_id") or "") not in by_id]
    for r in roots:
        is_root = r.get("proj_node_flag") == "Y"
        names = [] if is_root else [r.get("wbs_name", "")]
        code = "" if is_root else "1"
        out[r["wbs_id"]] = {"names": names, "code": code}
        walk(r["wbs_id"], names, code)
    return out


def import_xer(text: str) -> dict:
    tables = parse_tables(text)
    tasks = tables.get("TASK") or []
    if not tasks:
        raise ValueError("No TASK table found — this does not look like a P6 XER file")

    wbs = _wbs_paths(tables.get("PROJWBS") or [])

    used, id_by_task = set(), {}
    activities = []
    for t in tasks:
        code = (t.get("task_code") or "").strip() or f"A{len(activities) + 1:04d}"
        while code in used:
            code = f"{code}_1"
        used.add(code)
        id_by_task[t.get("task_id")] = code
        path = wbs.get(t.get("wbs_id"), {"names": [], "code": ""})
        names = path["names"]
        atype = TYPE_MAP.get((t.get("task_type") or "").strip(), "Task")
        duration = 0 if atype == "Milestone" else _hours_to_days(t.get("target_drtn_hr_cnt"))
        activities.append({
            "activity_id": code,
            "wbs_code": path["code"],
            "wbs_l1": names[0] if len(names) > 0 else "Imported",
            "wbs_l2": names[1] if len(names) > 1 else "",
            "wbs_l3": " / ".join(names[2:]) if len(names) > 2 else "",
            "description": (t.get("task_name") or code).strip(),
            "type": atype,
            "duration": duration,
            "predecessors": [],
            "constraint_type": CSTR_MAP.get((t.get("cstr_type") or "").strip(), ""),
            "constraint_date": _date(t.get("cstr_date")),
            "_start": _date(t.get("target_start_date")) or _date(t.get("early_start_date")),
        })

    by_code = {a["activity_id"]: a for a in activities}
    for link in tables.get("TASKPRED") or []:
        succ = id_by_task.get(link.get("task_id"))
        pred = id_by_task.get(link.get("pred_task_id"))
        if not succ or not pred or succ == pred:
            continue
        by_code[succ]["predecessors"].append({
            "id": pred,
            "type": PRED_MAP.get((link.get("pred_type") or "").strip(), "FS"),
            "lag": _hours_to_days(link.get("lag_hr_cnt")),
        })

    project_rows = tables.get("PROJECT") or []
    start = None
    if project_rows:
        p = project_rows[0]
        start = _date(p.get("plan_start_date")) or _date(p.get("scd_end_date"))
    if not start:
        starts = [a["_start"] for a in activities if a["_start"]]
        start = min(starts) if starts else date.today().isoformat()

    name = ""
    if project_rows:
        name = (project_rows[0].get("proj_short_name") or "").strip()

    week_days = 5
    holidays = []
    for c in tables.get("CALENDAR") or []:
        data = c.get("clndr_data") or ""
        working = len(re.findall(r"\(0\|\|\d\(\)\(0\|\|0", data))
        if working:
            week_days = min(7, max(1, working))
        exc = data.split("Exceptions", 1)
        if len(exc) > 1:
            holidays = sorted(
                {
                    f"{m[0:4]}-{m[4:6]}-{m[6:8]}"
                    for m in re.findall(r"\b(\d{8})\b", exc[1])
                    if 1990 <= int(m[0:4]) <= 2100 and 1 <= int(m[4:6]) <= 12 and 1 <= int(m[6:8]) <= 31
                }
            )
        if working or holidays:
            break

    for a in activities:
        a.pop("_start", None)

    return {
        "name": name or "Imported programme",
        "start_date": start,
        "week_pattern": {6: "6-day", 7: "7-day"}.get(week_days, "5-day"),
        "holidays": holidays,
        "activities": activities,
        "stats": {
            "activities": len(activities),
            "links": sum(len(a["predecessors"]) for a in activities),
            "milestones": sum(1 for a in activities if a["type"] == "Milestone"),
            "wbs_nodes": len(wbs),
            "holidays": len(holidays),
        },
    }

File: server/xerImport.ts
const TYPE_MAP: Record<string, "Milestone" | "Summary" | "Task"> = {
  TT_Mile: "Milestone",
  TT_FinMile: "Milestone",
  TT_StartMile: "Milestone",
  TT_WBS: "Summary",
  TT_LOE: "Summary",
  TT_Task: "Task",
  TT_Rsrc: "Task",
};

const PRED_MAP: Record<string, "FS" | "SS" | "FF" | "SF"> = {
  PR_FS: "FS",
  PR_SS: "SS",
  PR_FF: "FF",
  PR_SF: "SF",
};

const CSTR_MAP: Record<string, "" | "SNET" | "FNLT" | "MSO"> = {
  CS_MSO: "SNET",
  CS_MSOB: "SNET",
  CS_ALAP: "",
  CS_MANDSTART: "MSO",
  CS_MEO: "FNLT",
  CS_MEOB: "FNLT",
  CS_MANDFIN: "FNLT",
  CS_MSOA: "SNET",
  CS_MEOA: "FNLT",
};

export function parseTables(text: string): Record<string, Record<string, string>[]> {
  const tables: Record<string, Record<string, string>[]> = {};
  let name: string | null = null;
  let fields: string[] = [];

  const lines = text.split(/\r?\n/);
  for (const raw of lines) {
    if (!raw || raw.startsWith("ERMHDR")) continue;
    const parts = raw.split("\t");
    const tag = parts[0];
    if (tag === "%T") {
      name = parts[1]?.trim() || "";
      tables[name] = [];
      fields = [];
    } else if (tag === "%F") {
      fields = parts.slice(1).map((p) => p.trim());
    } else if (tag === "%R" && name && fields.length > 0) {
      const values = parts.slice(1);
      const row: Record<string, string> = {};
      for (let i = 0; i < fields.length; i++) {
        row[fields[i]] = i < values.length ? values[i] : "";
      }
      tables[name].push(row);
    } else if (tag === "%E") {
      break;
    }
  }
  return tables;
}

function parseDate(value: string | null | undefined): string | null {
  if (!value) return null;
  const m = value.trim().match(/(\d{4})-(\d{2})-(\d{2})/);
  if (!m) return null;
  return `${m[1]}-${m[2]}-${m[3]}`;
}

function hoursToDays(value: any): number {
  try {
    const num = parseFloat(value || "0");
    return Math.max(0, Math.round(num / 8.0));
  } catch {
    return 0;
  }
}

function buildWbsPaths(rows: Record<string, string>[] = []): Record<string, { names: string[]; code: string }> {
  const byId: Record<string, Record<string, string>> = {};
  const children: Record<string, Record<string, string>[]> = {};

  for (const r of rows) {
    byId[r.wbs_id] = r;
    const parentId = r.parent_wbs_id || "";
    if (!children[parentId]) children[parentId] = [];
    children[parentId].push(r);
  }

  for (const parentId in children) {
    children[parentId].sort((a, b) => {
      const seqA = parseInt(a.seq_num || "0", 10) || 0;
      const seqB = parseInt(b.seq_num || "0", 10) || 0;
      if (seqA !== seqB) return seqA - seqB;
      return (a.wbs_name || "").localeCompare(b.wbs_name || "");
    });
  }

  const out: Record<string, { names: string[]; code: string }> = {};

  function walk(nodeId: string, names: string[], code: string) {
    const kids = children[nodeId] || [];
    kids.forEach((kid, idx) => {
      const isRoot = kid.proj_node_flag === "Y";
      const kidNames = isRoot ? names : [...names, kid.wbs_name || ""];
      const num = idx + 1;
      const kidCode = isRoot ? code : code ? `${code}.${num}` : String(num);
      out[kid.wbs_id] = { names: kidNames, code: kidCode };
      walk(kid.wbs_id, kidNames, kidCode);
    });
  }

  const roots = rows.filter((r) => !r.parent_wbs_id || !byId[r.parent_wbs_id]);
  for (const r of roots) {
    const isRoot = r.proj_node_flag === "Y";
    const names = isRoot ? [] : [r.wbs_name || ""];
    const code = isRoot ? "" : "1";
    out[r.wbs_id] = { names, code };
    walk(r.wbs_id, names, code);
  }

  return out;
}

export function importXer(text: string): any {
  const tables = parseTables(text);
  const tasks = tables["TASK"] || [];
  if (tasks.length === 0) {
    throw new Error("No TASK table found — this does not look like a P6 XER file");
  }

  const wbs = buildWbsPaths(tables["PROJWBS"] || []);
  const used = new Set<string>();
  const idByTask: Record<string, string> = {};
  const activities: any[] = [];

  for (const t of tasks) {
    let code = (t.task_code || "").trim() || `A${String(activities.length + 1).padStart(4, "0")}`;
    while (used.has(code)) {
      code = `${code}_1`;
    }
    used.add(code);
    idByTask[t.task_id] = code;

    const path = wbs[t.wbs_id] || { names: [], code: "" };
    const names = path.names;
    const atype = TYPE_MAP[(t.task_type || "").trim()] || "Task";
    const duration = atype === "Milestone" ? 0 : hoursToDays(t.target_drtn_hr_cnt);

    activities.push({
      activity_id: code,
      wbs_code: path.code,
      wbs_l1: names.length > 0 ? names[0] : "Imported",
      wbs_l2: names.length > 1 ? names[1] : "",
      wbs_l3: names.length > 2 ? names.slice(2).join(" / ") : "",
      description: (t.task_name || code).trim(),
      type: atype,
      duration,
      predecessors: [],
      constraint_type: CSTR_MAP[(t.cstr_type || "").trim()] || "",
      constraint_date: parseDate(t.cstr_date),
      _start: parseDate(t.target_start_date) || parseDate(t.early_start_date),
    });
  }

  const byCode: Record<string, any> = {};
  for (const a of activities) {
    byCode[a.activity_id] = a;
  }

  for (const link of tables["TASKPRED"] || []) {
    const succ = idByTask[link.task_id];
    const pred = idByTask[link.pred_task_id];
    if (!succ || !pred || succ === pred || !byCode[succ]) continue;

    byCode[succ].predecessors.push({
      id: pred,
      type: PRED_MAP[(link.pred_type || "").trim()] || "FS",
      lag: hoursToDays(link.lag_hr_cnt),
    });
  }

  const projectRows = tables["PROJECT"] || [];
  let start: string | null = null;
  if (projectRows.length > 0) {
    const p = projectRows[0];
    start = parseDate(p.plan_start_date) || parseDate(p.scd_end_date);
  }
  if (!start) {
    const starts = activities.map((a) => a._start).filter(Boolean);
    start = starts.length > 0 ? starts.sort()[0] : new Date().toISOString().slice(0, 10);
  }

  let name = "";
  if (projectRows.length > 0) {
    name = (projectRows[0].proj_short_name || "").trim();
  }

  let weekDays = 5;
  let holidays: string[] = [];
  for (const c of tables["CALENDAR"] || []) {
    const data = c.clndr_data || "";
    const workingMatches = data.match(/\(0\|\|\d\(\)\(0\|\|0/g);
    if (workingMatches && workingMatches.length > 0) {
      weekDays = Math.min(7, Math.max(1, workingMatches.length));
    }
    const exc = data.split("Exceptions");
    if (exc.length > 1) {
      const hMatches = exc[1].match(/\b(\d{8})\b/g) || [];
      const set = new Set<string>();
      for (const m of hMatches) {
        const y = parseInt(m.slice(0, 4), 10);
        const mon = parseInt(m.slice(4, 6), 10);
        const day = parseInt(m.slice(6, 8), 10);
        if (y >= 1990 && y <= 2100 && mon >= 1 && mon <= 12 && day >= 1 && day <= 31) {
          set.add(`${m.slice(0, 4)}-${m.slice(4, 6)}-${m.slice(6, 8)}`);
        }
      }
      holidays = Array.from(set).sort();
    }
    if (workingMatches || holidays.length > 0) {
      break;
    }
  }

  for (const a of activities) {
    delete a._start;
  }

  return {
    name: name || "Imported programme",
    start_date: start,
    week_pattern: weekDays === 6 ? "6-day" : weekDays === 7 ? "7-day" : "5-day",
    holidays,
    activities,
    stats: {
      activities: activities.length,
      links: activities.reduce((sum, a) => sum + (a.predecessors?.length || 0), 0),
      milestones: activities.filter((a) => a.type === "Milestone").length,
      wbs_nodes: Object.keys(wbs).length,
      holidays: holidays.length,
    },
  };
}
